staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/controller.go

```diff
@@ -164,76 +199,62 @@ func (c *Controller) sync(name string) error {
 
 	// do we have to remove all specs of this CRD?
 	if errors.IsNotFound(err) || !apiextensionshelpers.IsCRDConditionTrue(crd, apiextensionsv1.Established) {
-		if _, found := c.crdSpecs[name]; !found {
+		if _, found := c.specsByName[name]; !found {
 			return nil
 		}
-		delete(c.crdSpecs, name)
+		delete(c.specsByName, name)
 		klog.V(2).Infof("Updating CRD OpenAPI spec because %s was removed", name)
 		regenerationCounter.With(map[string]string{"crd": name, "reason": "remove"})
-		return c.updateSpecLocked()
+		c.updateSpecLocked()
+		return nil
 	}
 
-	// compute CRD spec and see whether it changed
-	oldSpecs, updated := c.crdSpecs[crd.Name]
-	newSpecs, changed, err := buildVersionSpecs(crd, oldSpecs)
-	if err != nil {
-		return err
-	}
-	if !changed {
+	// If CRD spec already exists, update the CRD.
+	// updateSpecCacheCRD includes the ETag so an update on a spec
+	// resulting in the same ETag will be a noop.
+	s, exists := c.specsByName[crd.Name]
+	if exists {
+		s.update(crd)
 		return nil
 	}
 
 	// update specs of this CRD
-	c.crdSpecs[crd.Name] = newSpecs
+	c.specsByName[crd.Name] = createSpecCache(crd)
 	klog.V(2).Infof("Updating CRD OpenAPI spec because %s changed", name)
 	reason := "add"
-	if updated {
+	if exists {
```

Comment:
this is always false

Reply:
Restructured this section a bit. I feel like the metrics have less value now because updates aren't guaranteed for a reaggregation until the next request comes in.